File: packages/openmemory-js/src/server/routes/langgraph.ts
import { z } from "zod";
import {
    store_node_mem,
    retrieve_node_mems,
    get_graph_ctx,
    create_refl,
    get_lg_cfg,
} from "../../ai/graph";
import { require_tenant, reject_tenant_mismatch } from "../middleware/tenant";
import type {
    lgm_store_req,
    lgm_retrieve_req,
    lgm_context_req,
    lgm_reflection_req,
} from "../../core/types";

const LgmStoreReqSchema = z
    .object({
        memory: z.record(z.any()).optional(),
        content: z.string().optional(),
        metadata: z.record(z.any()).optional(),
        session_id: z.string().optional(),
        tags: z.array(z.string()).optional(),
    })
    .passthrough();

const LgmRetrieveReqSchema = z
    .object({
        query: z.string().optional(),
        session_id: z.string().optional(),
        limit: z.number().optional(),
    })
    .passthrough();

const LgmContextReqSchema = z
    .object({
        session_id: z.string().optional(),
    })
    .passthrough();

const LgmReflectionReqSchema = z
    .object({
        session_id: z.string().optional(),
    })
    .passthrough();

export function lg(app: any) {
    app.get("/lgm/config", (_req: any, res: any) => {
        res.json(get_lg_cfg());
    });

    app.post("/lgm/store", async (req: any, res: any) => {
        const tenant = require_tenant(req, res);
        if (!tenant) return;
        try {
            const parsedBody = LgmStoreReqSchema.parse(req.body);
            if (reject_tenant_mismatch(res, tenant, parsedBody.user_id)) return;
            const r = await store_node_mem({
                ...parsedBody,
                user_id: tenant,
            } as lgm_store_req);
            res.json(r);
        } catch (e) {
            console.error("[LGM] store error:", e);
            res.status(400).json({
                err: "lgm_store_failed",
                message: "internal",
            });
        }
    });

    app.post("/lgm/retrieve", async (req: any, res: any) => {
        const tenant = require_tenant(req, res);
        if (!tenant) return;
        try {
            const parsedBody = LgmRetrieveReqSchema.parse(req.body);
            if (reject_tenant_mismatch(res, tenant, parsedBody.user_id)) return;
            const r = await retrieve_node_mems({
                ...parsedBody,
                user_id: tenant,
            } as lgm_retrieve_req);
            res.json(r);
        } catch (e) {
            console.error("[LGM] retrieve error:", e);
            res.status(400).json({
                err: "lgm_retrieve_failed",
                message: "internal",
            });
        }
    });

    app.post("/lgm/context", async (req: any, res: any) => {
        const tenant = require_tenant(req, res);
        if (!tenant) return;
        try {
            const parsedBody = LgmContextReqSchema.parse(req.body);
            if (reject_tenant_mismatch(res, tenant, parsedBody.user_id)) return;
            const r = await get_graph_ctx({
                ...parsedBody,
                user_id: tenant,
            } as lgm_context_req);
            res.json(r);
        } catch (e) {
            console.error("[LGM] context error:", e);
            res.status(400).json({
                err: "lgm_context_failed",
                message: "internal",
            });
        }
    });

    app.post("/lgm/reflection", async (req: any, res: any) => {
        const tenant = require_tenant(req, res);
        if (!tenant) return;
        try {
            const parsedBody = LgmReflectionReqSchema.parse(req.body);
            if (reject_tenant_mismatch(res, tenant, parsedBody.user_id)) return;
            const r = await create_refl({
                ...parsedBody,
                user_id: tenant,
            } as lgm_reflection_req);
            res.json(r);
        } catch (e) {
            console.error("[LGM] reflection error:", e);
            res.status(400).json({
                err: "lgm_reflection_failed",
                message: "internal",
            });
        }
    });
}

File: packages/openmemory-py/src/openmemory/server/routers/langgraph.py
from fastapi import APIRouter, HTTPException, Depends, Body
from typing import Dict, Any, List, Optional
from pydantic import BaseModel
from ...main import Memory
from ...core.types import LgStoreResult, LgRetrieveResult, LgContextResult, LgReflectResult
from ...ai.graph import (  # type: ignore[attr-defined]
    store_node_mem,
    LgmStoreReq,
    LgmContextReq,
    LgmRetrieveReq,
    LgmReflectionReq,
)

router = APIRouter(tags=["langgraph"])

@router.get("/config")
async def get_lg_config(memory: Memory = Depends(lambda: Memory())):
    """Get LGM Config."""
    # Assuming helper exists in graph.py (added in Phase 14)
    from ...ai.graph import get_lg_cfg  # type: ignore[attr-defined]

    return get_lg_cfg()

@router.post("/store", response_model=LgStoreResult)
async def lg_store(req: LgmStoreReq):
    """Store LGM memory."""
    res = await store_node_mem(req)
    return res

@router.post("/retrieve", response_model=LgRetrieveResult)
async def lg_retrieve(req: LgmRetrieveReq):
    """Retrieve LGM memories."""
    from ...ai.graph import retrieve_node_mems  # type: ignore[attr-defined]

    res = await retrieve_node_mems(req)
    return res

@router.post("/context", response_model=LgContextResult)
async def lg_context(req: LgmContextReq):
    """Get distilled context."""
    from ...ai.graph import get_graph_ctx  # type: ignore[attr-defined]

    res = await get_graph_ctx(req)
    return res

@router.post("/reflection", response_model=LgReflectResult)
async def lg_reflect(req: LgmReflectionReq):
    """Trigger reflection."""
    from ...ai.graph import create_refl  # type: ignore[attr-defined]

    res = await create_refl(req)
    return res
